File: recognition/v1/python3/recognize.py
#!/usr/bin/env python3

import argparse
import itertools
import time

import grpc

import recognition_pb2
import recognition_pb2_grpc


CHUNK_SIZE = 2048
SLEEP_TIME = 0.1

ENCODING_PCM = 'pcm'

ENCODINGS_MAP = {
    ENCODING_PCM: recognition_pb2.RecognitionOptions.PCM_S16LE,
    'opus': recognition_pb2.RecognitionOptions.OPUS,
    'mp3': recognition_pb2.RecognitionOptions.MP3,
    'flac': recognition_pb2.RecognitionOptions.FLAC,
    'alaw': recognition_pb2.RecognitionOptions.ALAW,
    'mulaw': recognition_pb2.RecognitionOptions.MULAW,
}


def try_printing_request_id(md):
    for m in md:
        if m.key == 'x-request-id':
            print('RequestID:', m.value)


def generate_audio_chunks(path, chunk_size=CHUNK_SIZE, sleep_time=SLEEP_TIME):
    with open(path, 'rb') as f:
        for data in iter(lambda: f.read(chunk_size), b''):
            yield recognition_pb2.RecognitionRequest(audio_chunk=data)
            time.sleep(sleep_time)


def recognize(args):
    ssl_cred = grpc.ssl_channel_credentials()
    token_cred = grpc.access_token_call_credentials(args.token)

    channel = grpc.secure_channel(
        args.host,
        grpc.composite_channel_credentials(ssl_cred, token_cred),
    )

    stub = recognition_pb2_grpc.SmartSpeechStub(channel)

    con = stub.Recognize(itertools.chain(
        (recognition_pb2.RecognitionRequest(options=args.recognition_options),),
        generate_audio_chunks(args.file),
    ))

    try:
        for resp in con:
            if not resp.eou:
                print('Got partial result:')
            else:
                print('Got end-of-utterance result:')

            for i, hyp in enumerate(resp.results):
                print('  Hyp #{}: {}'.format(i + 1, hyp.normalized_text if args.normalized_result else hyp.text))

            if resp.eou and args.emotions_result:
                print('  Emotions: pos={}, neu={}, neg={}'.format(
                    resp.emotions_result.positive,
                    resp.emotions_result.neutral,
                    resp.emotions_result.negative,
                ))
    except grpc.RpcError as err:
        print('RPC error: code = {}, details = {}'.format(err.code(), err.details()))
    except Exception as exc:
        print('Exception:', exc)
    else:
        print('Recognition has finished')
    finally:
        try_printing_request_id(con.initial_metadata())
        channel.close()


class Arguments:
    NOT_RECOGNITION_OPTIONS = {'host', 'token', 'file', 'normalized_result', 'emotions_result'}
    NOT_RECOGNITION_OPTIONS.update({'ca', 'cert', 'key'})  # diff
    DURATIONS = {'no_speech_timeout', 'max_speech_timeout', 'eou_timeout'}
    REPEATED = {'words', 'insight_models'}
    HINTS_PREFIX = 'hints_'
    SPEAKER_SEPARATION_PREFIX = 'speaker_separation_options_'

    def __init__(self):
        super().__setattr__('recognition_options', recognition_pb2.RecognitionOptions())

    def __setattr__(self, key, value):
        if key in self.NOT_RECOGNITION_OPTIONS:
            super().__setattr__(key, value)
        elif key.startswith(self.HINTS_PREFIX):
            key = key[len(self.HINTS_PREFIX):]
            self._set_option(self.recognition_options.hints, key, value)
        elif key.startswith(self.SPEAKER_SEPARATION_PREFIX):
            key = key[len(self.SPEAKER_SEPARATION_PREFIX):]
            self._set_option(self.recognition_options.speaker_separation_options, key, value)
        else:
            self._set_option(self.recognition_options, key, value)

    def _set_option(self, obj, key, value):
        if key in self.DURATIONS:
            getattr(obj, key).FromJsonString(value)
        elif key in self.REPEATED:
            if value:
                getattr(obj, key).extend(value)
        else:
            setattr(obj, key, value)


def create_parser():
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    parser.add_argument('--host', default='smartspeech.sber.ru', help='host:port of gRPC endpoint')
    parser.add_argument('--token', required=True, default=argparse.SUPPRESS, help='access token')
    parser.add_argument('--file', required=True, default=argparse.SUPPRESS, help='audio file for recognition')

    parser.add_argument('--normalized-result', action='store_true', help='show normalized text')
    parser.add_argument('--emotions-result', action='store_true', help='show emotions result')

    parser.add_argument('--audio-encoding', default=ENCODINGS_MAP[ENCODING_PCM], type=lambda x: ENCODINGS_MAP[x], help=str(list(ENCODINGS_MAP)))
    parser.add_argument('--sample-rate', default=16000, type=int, help='PCM only')
    parser.add_argument('--model', default='', help=' ')
    parser.add_argument('--hypotheses-count', default=1, type=int, help=' ')
    parser.add_argument('--enable-profanity-filter', action='store_true', help=' ')
    parser.add_argument('--enable-multi-utterance', action='store_true', help=' ')
    parser.add_argument('--enable-partial-results', action='store_true', help=' ')
    parser.add_argument('--no-speech-timeout', default='7s', help=' ')
    parser.add_argument('--max-speech-timeout', default='20s', help=' ')
    parser.add_argument('--hints-words', nargs='*', default=[], help=' ')
    parser.add_argument('--hints-enable-letters', action='store_true', help=' ')
    parser.add_argument('--hints-eou-timeout', default='0s', help=' ')
    parser.add_argument('--channels-count', default=1, type=int, help=' ')
    parser.add_argument('--speaker-separation-options-enable', action='store_true', help=' ')
    parser.add_argument('--speaker-separation-options-enable-only-main-speaker', action='store_true', help=' ')
    parser.add_argument('--speaker-separation-options-count', default=0, type=int, help=' ')
    parser.add_argument('--insight-models', nargs='*', default=[], help=' ')

    return parser


def main():
    parser = create_parser()

    recognize(parser.parse_args(namespace=Arguments()))


if __name__ == '__main__':
    main()

File: recognition/v1/nodejs/recognize.js
#!/usr/bin/env node

const fs = require('fs');
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');

const HOST = 'smartspeech.sber.ru';
const OPTIONS = {
    language: 'ru-RU',
    model: 'general',
    hypotheses_count: 1,
    enable_partial_results: false,
    enable_multi_utterance: false,
    enable_profanity_filter: false,
    no_speech_timeout: {seconds: 7},
    max_speech_timeout: {seconds: 20},
    hints: {
        words: [],
        enable_letters: false,
        hints_eou_timeout: {seconds: 0},
    },
};
const PROTO = __dirname + '/../recognition.proto';

function recognize(accessToken, file, sampleRate) {
    const packageDefinition = protoLoader.loadSync(PROTO, {
        keepCase: true, longs: String, enums: String, defaults: true, oneofs: true, includeDirs: [__dirname + '/../../../task/v1/']});
    const proto = grpc.loadPackageDefinition(packageDefinition);

    const service = new proto['smartspeech']['recognition']['v1']['SmartSpeech'](HOST, grpc.credentials.createSsl());

    let metadata = new grpc.Metadata();
    metadata.set('authorization', 'Bearer ' + accessToken);

    const call = service['Recognize'](metadata);
    call.on('metadata', (metadata) => {
        console.info('X-Request-Id:', metadata.get('x-request-id'));
    });
    call.on('data', (response) => {
        console.info('Response:')
        if (response.results && response.results.length > 0) {
            console.info(`  Raw text: ${response.results[0].text}`);
            console.info(`  Normalized text: ${response.results[0].normalized_text}`);
            console.info(`  Is final: ${response.eou}`);
        }
    });
    call.on('error', (error) => {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    });
    call.on('end', () => {
        call.end(function () {
            service.close();
            process.exit(0);
        });
    });

    const deadline = new Date(Date.now() + 3000);
    service.waitForReady(deadline, (error) => {
        if (!error) {
            const options = Object.assign({
                audio_encoding: (/\.ogg|.opus$/.test(file)) ? 'OPUS' : 'PCM_S16LE',
                sample_rate: sampleRate,
            }, OPTIONS);
            call.write({options: options});

            const reader = fs.createReadStream(file, {highWaterMark: 3200});
            reader.on('data', (chunk) => {
                call.write({audio_chunk: chunk})
            });
            reader.on('error', (error) => {
                console.error('Failed to read file:', error.message);
                process.exit(1);
            });
            reader.on('end', () => {
                call.end()
            });
        } else {
            console.error(`Failed to connect to ${HOST}:`, error);
            process.exit(1);
        }
    });
}

function main() {
    try {
        fs.lstatSync(PROTO).isFile();
    } catch (error) {
        console.error(`Failed to load proto file ${PROTO}:`, error);
        process.exit(1);
    }

    const argv = process.argv.slice(1);
    if (argv.length < 3 || argv.length > 4) {
        console.error(`usage: node ${argv[0]} <access_token> <file> [<sample_rate>]`);
        process.exit(1);
    }
    const accessToken = argv[1];
    const file = argv[2];
    const sampleRate = argv[3] || 16000;
    recognize(accessToken, file, sampleRate)
}

main()
